<?xml version="1.0" encoding="UTF-8"?>
<launch>
  <arg name="model" default="$(find SRBProject)/robots/slab_srb.gazebo.xacro" />
  <arg name="rvizconfig" default="$(find SRBProject)/rviz/slab_srb.rviz" />  
  
  <!-- gazebo world -->
  <include file="$(find gazebo_ros)/launch/empty_world.launch">
    <arg name="world_name" value="$(find SRBProject)/worlds/test1.world"/>
    <arg name="paused" value="false"/>
    <arg name="use_sim_time" value="true"/>
    <arg name="gui" value="true"/>
    <arg name="headless" value="false"/>
    <arg name="debug" value="false"/>
  </include>

  <!-- slab_srb_robot -->
  <include file="$(find SRBProject)/launch/fake_srb_spawn.launch">
    <arg name="robot_name" value="slab_srb_robot"/>
  </include>

  <group ns="slab_srb_robot">

    <!-- to increase the wheel odometry error -->
    <param name="/slab_srb_robot/diff_drive_controller/wheel_radius" value="0.1225" />

    <!-- visal_indicator -->
    <node pkg="slab_srb" type="srb_visual_indicator_pub" name="visual_indicator">
    </node>

    <!-- PS3joy -->
    <node pkg="joy" name="joy_node" type="joy_node">
    </node>

    <!-- joy_to_cmd_twist -->
    <node pkg="SRBProject" name="JoyToTwistPublisher" type="JoyToTwistPublisher">
    </node>

    <!-- laser_filter -->
    <node pkg="laser_filters" type="scan_to_scan_filter_chain" name="laser_filter">
        <param name="frame_id" value="slab_srb_robot/front_laser_link"/>
        <rosparam command="load" file="$(find SRBProject)/config/laser_filter/slab_srb_laser_filter.yaml" />
        <remap from="scan" to="front_laser/scan_raw" />
        <remap from="scan_filtered" to="front_laser/scan" />
    </node>

    <!-- imu_filter -->
    <node pkg="slab_srb" name="slab_srb_imu_data_filter" type="slab_srb_imu_data_filter">
    </node>

    <!-- robot localization -->
    <node pkg="robot_localization" type="ekf_localization_node" name="ekf_se" clear_params="true" output="screen">
      <rosparam command="load" file="$(find SRBProject)/config/ekf_localization/slab_srb_ekf.yaml" />
      <param name="tf_prefix" value="slab_srb_robot"/>
      <remap from="odometry/filtered" to="fusion/odom" />
      <remap from="odom0" to="diff_drive_controller/odom" />
      <remap from="imu0"  to="imu/data" />
    </node>

    <!-- gmapping -->
    <node pkg="gmapping" type="slam_gmapping" name="slam_gmapping" clear_params="true">
      <rosparam command="load" file="$(find SRBProject)/config/gmapping.yaml" />
      <remap from="scan" to="front_laser/scan" />
      <param name="base_frame" value="slab_srb_robot/base_link" />
      <param name="odom_frame" value="slab_srb_robot/odom" />
      <param name="map_frame"  value="slab_srb_robot/map" />
    </node>

  </group>

  <!-- rviz -->
  <node name="rviz" pkg="rviz" type="rviz" args="-d $(arg rvizconfig)" required="true" />
</launch>
